src/react/useMutation.ts

```diff
@@ -13,19 +11,64 @@ import {
   uid,
   noop,
 } from '../core/utils'
+import {
+  QueryStatus,
+  MutationResultPair,
+  MutationFunction,
+  MutationConfig,
+} from '../core/types'
+
+// TYPES
+
+type Reducer<S, A> = (prevState: S, action: A) => S
+
+interface State<TResult, TError> {
+  status: QueryStatus
+  data?: TResult
+  error: TError | null
+}
+
+interface ResetAction {
+  type: 'Reset'
+}
+
+interface LoadingAction {
+  type: 'Loading'
+}
+
+interface ResolveAction<TResult> {
+  type: 'Resolve'
+  data: TResult
+}
+
+interface RejectAction<TError> {
+  type: 'Reject'
+  error: TError
+}
 
-const getDefaultState = () => ({
+type Action<TResult, TError> =
+  | ResetAction
+  | LoadingAction
+  | ResolveAction<TResult>
+  | RejectAction<TError>
+
+// HOOK
+
+const getDefaultState = (): State<any, any> => ({
   status: statusIdle,
   data: undefined,
   error: null,
 })
 
-const actionReset = {}
-const actionLoading = {}
-const actionResolve = {}
-const actionReject = {}
+const actionReset = 'Reset'
+const actionLoading = 'Loading'
+const actionResolve = 'Resolve'
+const actionReject = 'Reject'
```

Comment:
Let's have an `enum` for these types and use them in above interfaces and in the below comparisons. It would help in consistency. It could be something like:
```ts
enum ActionType {
	RESET = 'Reset',
	LOADING = 'Loading',
	...
}

interface ResetAction {
	type: ActionType.RESET
}

...

if (action.type === ActionType.RESET) {
	...
}
```

Reply:
Good suggestion, done!